### ISSUE TITLE
Use a white list for domains able to authenticate in a multi-tenant application environment

### ISSUE THREAD

**Issue by evgcardoso** [wpo365/wpo365-login#16]
I am working on a multi-tenant application based on wordpress to offer to my customers and your plugin is handling the authentication part perfectly when using a multi-tenant app on AAD and setting the tenant ID to common on the plugin configuration. The only thing that is missing is the ability to define a white list of the domains I want to authorize to authenticate this way, so that I can prevent undesired users with a Office365 account signing-in to my application unless I authorize them first. Any user with an unauthorized domain would then be redirected back to Wordpress default login.php. I could achieve this by disabling automatic user creation, but then I have to create user accounts manually for each and every user and that is far from optimal.

I am looking at the code and I believe this would not be very hard to implement by myself, but if you would consider the possibility of adding this feature to a future release of the plugin it would be really great.

**Comment by wpo365** [comment]
I see where you coming from and I agree that this makes sense. I will update the plugin shortly. Having said that I found that another issue with this scenario is the fact that the WPO365 Options are visible to all admin (since the plugin must still be network activated). As an alternative, I've created for a project of my own a version where all settings are configured as global variables and can be added to global config. I guess I will add this option as well and I guess this will be a good thing for you too.

**Comment by wpo365** [comment]
@evgcardoso I have updated the plugin and added an option to white list domains. It compares a user's email (smtp) domain with entries in that list. So the plugin now works as follows:
1. No domains white listed => works as before i.e. all domains are accepted for user registration.
2. Domain(s) white listed => only users with email host found in the white list can register.
Let me know if this solutions fits your purpose.

**Comment by stevegroner** [comment]
So would this also handle restricting the application to specific groups in Azure AD?  Right now, when the person tries to authenticate when I have a SEC group assigned to the application in Azure AD, if they do not pass validation it send the user to the wp-login page, but with a bunch of error complaining about modifying headers that have already been sent.

We have multiple domains in the same tenant and I control application access via users/groups assigned to the application.

**Comment by wpo365** [comment]
Hi @stevegroner In version 3.5 the session validation will start as soon as possible. Ideally this means that in your case authentication happens before headers are already sent. Basically the error "cannot modify headers ..." is because the plugin "aborts" execution of the current request when it sends the current user to the login page. However, session validation should run before any response is being sent back to the user. But it may be that other plugins also try and launch as soon as possible and start writing back to the user. This is something that is difficult to prevent or orchestrate. But please have a go and install version 3.5 and let me know if this solved your issue.

**Comment by stevegroner** [comment]
Ok, I set (2) domains in the domain whitelist.  When I attempt to login with a user that is not one of those, it does not graceful handle that.  Here is a the text of the error.  The only plugins installed on this site are also listed below.

Warning: Cannot modify header information - headers already sent by (output started at /home/itcawgcap/public_html/wp-login.php:75) in /home/itcawgcap/public_html/wp-content/plugins/wpo365-login/Wpo/Util/Helpers.php on line 22

Warning: Cannot modify header information - headers already sent by (output started at /home/itcawgcap/public_html/wp-login.php:75) in /home/itcawgcap/public_html/wp-content/plugins/wpo365-login/Wpo/Util/Helpers.php on line 23

Warning: Cannot modify header information - headers already sent by (output started at /home/itcawgcap/public_html/wp-login.php:75) in /home/itcawgcap/public_html/wp-content/plugins/wpo365-login/Wpo/Util/Helpers.php on line 24

Could not create or retrieve your login. Please contact your System Administrator.

Plugin List:

- Display PHP Version
- JS Support Ticket
- Redux Framework
- WPO365-login

If I use a user account that is in the whitelist, it works successfully.

**Comment by wpo365** [comment]
@stevegroner Please try and update to version 3.6. I've updated the way the plugin now passes error information between pages and no cookies are used anymore. Therefore the warnings should no longer appear.

**Comment by stevegroner** [comment]
After installing 3.6 I cannot even open the site.  I receive this message immediately when the site tries to load.


Fatal error: Arrays are not allowed in class constants in /home/reservationscawg/public_html/wp-content/plugins/wpo365-login/Wpo/Util/Error_Handler.php on line 24

**Comment by stevegroner** [comment]
If you want I can privately send you some test logins to our tenant to play with.  Just let me know.

**Comment by wpo365** [comment]
@stevegroner many thanks for bringing this to my attention. I strongly suspect that this has to do with a somewhat older version you may be using of PHP. I've updated the plugin with version 3.7 and implemented a work-around.

**Comment by stevegroner** [comment]
Yes, I can confirm this works.  So it appears that perhaps your using session variables now?  Once the USER_NOT_FOUND , the only way to get rid of it, is to close all browser windows for that browser type, correct?

**Comment by stevegroner** [comment]
BTW, PHP version is 5.4.45

**Comment by wpo365** [comment]
It's not an environment variable but it's added as a wp query variable to the URL. So as long as you just refresh this page, the URL won't change and you'll see this (wp) query string variable. But it should not behave like a session variable. In other words, when you navigate away from the login page, the login_errors" query string variable should be gone (unless the same error occurs of course again). I guess, however, that in such a case a user will navigate away, because login was not successful. But let me know if you have a good idea to further optimize this.

**Comment by stevegroner** [comment]
Just for clarification, we have multiple domains under our tenant.  I am hitting the site using a login that is valid in our account, but does not meet the domain whitelist requirements.  Therefore is successfully authenticates the user and sets the local token.  Then it tries to check the domain whitelist.  When that fails which it should, when you navigate away and come back, the MSOL token for the previous user already exists and so it never asks again for a user, it just re-uses the existing token and immediately goes to the whitelist check which then fails as it should.  But there is no way to input a new user, unless you close the browser windows and start again, which invalidates the MSOL token.  Make sense?

**Comment by wpo365** [comment]
@stevegroner yes and no ... it's kind of a two-worlds thing: A user that successfully signed in with Microsoft may still not be able to authenticate with Wordpress. The Wordpress authentication always happen after the user returns from Microsoft (which will only happen when he/she successfully authenticated). The plugin, in that case, gets two pieces of information from Microsoft: 1. Login (with MSOL) was successful and 2. some basic user (identity) information. When if starts authenticating the user against the Wordpress user management module it may turn out that his login attempt should be rejected (e.g. because his/her domain is whitelisted).

At this moment it is not possible to sign-out a user from Microsoft (because the auth cookie is set by Microsoft). Technically, this could be implemented using a logout page (I'll keep this in mind). This similar behaviour you'll see when you sign into O365 with Account "A". To logout, because you want to sign in with Account "B" you must actively log out from Microsoft.

Ps, You may notice that when you want to logon with a different user, you're still being redirected to Microsoft. However, you're being redirected almost immediately, because it recognizes, that you're already signed in.

Tip: For testing purposes it makes sense to use the Google Chrome browser and create multiple users. Each user will then "remember" its own set of credentials and you don't need to clear the cache.

**Comment by stevegroner** [comment]
Sure, I understand.  Chicken or the egg situation.  You do not know their actual domain until they are authenticated, which is being passed into the system.  Perhaps my other suggestion will resolve this issue.  I posted another issue (enhancement request).  Perhaps that will help to address this issue.  Not sure how deep you want to go with this plugin, but I am simply making the request as it would certainly help me.  Thanks for the hard work, its an excellent plugin.

**Comment by wpo365** [comment]
Thanks for the input adn feedback @stevegroner  Would be great if you would recommend it to others over at the Wordpress plugin directory https://wordpress.org/plugins/search/wpo365-login/ ...

**Comment by stevegroner** [comment]
Done!  Awesome...
